modules/bash-commons/src/aws.sh

```diff
@@ -0,0 +1,80 @@
+#!/bin/bash
+# A collection of thin wrappers for direct calls to the AWS CLI and EC2 metadata API. These wrappers exist so that
+# (a) it's more convenient to fetch specific info you need, such as an EC2 Instance's private IP and (b) so you can
+# replace these helpers with mocks to do local testing or unit testing.
+
+set -e
+
+# Look up the given path in the EC2 Instance metadata endpoint
+function aws_lookup_path_in_instance_metadata {
+  local readonly path="$1"
+  curl --silent --show-error --location "http://169.254.169.254/latest/meta-data/$path/"
+}
+
+# Look up the given path in the EC2 Instance dynamic metadata endpoint
+function aws_lookup_path_in_instance_dynamic_data {
+  local readonly path="$1"
+  curl --silent --show-error --location "http://169.254.169.254/latest/dynamic/$path/"
+}
+
+# Get the private IP address for this EC2 Instance
+function aws_get_instance_private_ip {
+  aws_lookup_path_in_instance_metadata "local-ipv4"
+}
+
+# Get the public IP address for this EC2 Instance
+function aws_get_instance_public_ip {
+  aws_lookup_path_in_instance_metadata "public-ipv4"
+}
```

Comment:
This functions needs to be updated per https://github.com/gruntwork-io/package-kafka/commit/9ecc64e8230feb5239a86faa3e18908ac715dd4f and https://github.com/gruntwork-io/package-kafka/commit/b526ee3f445f55cf93e4bd744adc0ff98c2e3b86.

Reply:
Hm, I'm not convinced falling back magically to a private IP when someone requested a public IP is the right behavior. In most cases, this error implies something is wrong (e.g., you forgot to enable public IPs for your EC2 Instance), and swallowing that error will just lead to other errors, but somewhere far away in the code, where it'll be much more confusing.

If we want fall back behavior, we should add a separate method dedicated to that.

Reply:
The use case that motivated this was that [this line](https://github.com/gruntwork-io/package-kafka/blob/master/modules/run-kafka/bin/run-kafka#L572) was failing because there was no public IP defined for the box. The public-ip value is needed for [this line](https://github.com/gruntwork-io/package-kafka/blob/master/modules/run-kafka/bin/run-kafka#L587), which is used to process the Kafka "advertised listeners" property (e.g. `EXTERNAL://<public-ip>` should be dynamically updated to `EXTERNAL://1.2.34` at run-time). 

I suppose I could update the code to only attempt to replace vars that are actually present. This way, if the user specifies `<public-ip>` but in fact no such value is available, they'll get a well-defined error.

Thoughts?

Reply:
> I suppose I could update the code to only attempt to replace vars that are actually present. This way, if the user specifies <public-ip> but in fact no such value is available, they'll get a well-defined error.

Yup, that seems like the best option. I suspect there may be other "replacement vars" in the future that should only be computed if requested, and it would be an error to try to compute them otherwise. So it would be good to have conditional replacement. 

BTW, I used a slightly different pattern with Couchbase: https://github.com/gruntwork-io/terraform-aws-couchbase/tree/master/modules/run-sync-gateway#quick-start. Instead of magically replacing all placeholders in the config file, I allow the user to specify the name of the placeholder to replace. Example usage: https://github.com/gruntwork-io/terraform-aws-couchbase/blob/master/examples/couchbase-multi-cluster/user-data/user-data-sync-gateway.sh#L20-L28.

Reply:
Ok, I've noted that down as a ToDo in the next PR for Kafka. Once the `aws.sh` script looks good there, I'll open a PR here. The basic change is to have a nice error message (not `curl (22) Failed to find URL`).

On the replacement vars, I like the format you show because it's very obvious what's happening. Where it gets tricky is when you have a need for more "complex" vars. For example, Kafka has [4 different complex vars](https://github.com/gruntwork-io/package-kafka/tree/master/modules/run-kafka/config#available-replacement-variables), so in the case of Kafka, you're either looking up which vars are pre-defined, or which args you need to pass that do fancy replacement.